src/tools/grpc_server.cc

```diff
@@ -159,9 +162,31 @@ Status parse_query(query *q, const ::Query* request, ::CodeSearchResult* respons
 
 class add_match {
 public:
-    add_match(CodeSearchResult* response) : response_(response) {}
+    typedef std::set<std::string> line_set;
+
+    add_match(CodeSearchResult* response)
+        : unique_lines_(NULL), response_(response) {}
+
+    add_match(line_set* unique_lines, CodeSearchResult* response)
+        : unique_lines_(unique_lines), response_(response) {}
+
+    int match_count() {
+        return response_->results_size();
+    }
 
     void operator()(const match_result *m) const {
+        if (unique_lines_ != NULL) {
+            // Avoid a duplicate if a line is returned once from the
+            // tags search then again during the main corpus search.
+            std::string key = std::string(m->file->tree->name)
```

Comment:
It's prooobably fine here but doing a lot of string concatenation makes me kind of sad; it's got poor performance (probably fine in practice since these things are mostly short), and it could go wrong if `name` or `version` have spaces.

Let's instead store a `std::pair<indexed_file*,int>`, which should uniquely identify lines (there's only one `indexed_file` per file, so comparing by address is fine). That will be both shorter, faster, and more accurate.

Reply:
Especially if we make that change, this should be ~free both in terms of performance and in terms of having no impact in the normal case, so I'd make this deduplication unconditional, store just a `std::set` instead of a pointer to one, and simplify the code thereby.

Reply:
:+1:

Reply:
@nelhage Attempting to store the `std::set` on the class itself runs afoul of the `const` declaration of `void operator()(const match_result *m) const {`. The `const` seems to be there because the callback will later be coerced to a `std::function<void (const struct match_result*)>` which, I am told by @jboning, makes a copy of the whole struct, and it would be the copy’s `set` that received the lines instead of the original struct's `set`.

I'll go ahead and attempt to improve the type of the set so we can avoid the string concatenation, though!

Reply:
Ah yeah, that makes sense, I forgot that we were passing these by value. I think storing a reference or pointer to external storage makes sense then. But yeah, let's still improve the type.